### PR TITLE
Adding new RankFeature phase

### PR DESCRIPTION
**Pull request by pmpailis**
In this PR we add a new search phase, in-between `query` and `fetch`, that is responsible for applying any reranking needed. 

The idea is to trim down the query phase results down to `window_size`, reach out to the shards to extract any additional feature data if needed, and then use this information to rerank the top results, trim them down to `size` and pass them to `fetch` phase. 

In this "introductory" PR, the only operation that takes place is the reduction of the results through the `org.elasticsearch.action.search.QueryPhaseResultConsumer#reduce`, which was previously happening in the `FetchSearchPhase`.  This is required because we need to identify and operate on just the top results from all shard responses.

Moving the `reduce` operation could have subtle side-effects, so this is open for discussion, i.e. whether this should take place in this new phase, or if we'd need to re-design the approach. 

Relates to https://github.com/elastic/elasticsearch/issues/105472

### CODE REVIEW

**Comment by pmpailis** [comment]
run elasticsearch-ci/part-1

File: server/src/main/java/org/elasticsearch/action/search/FetchSearchPhase.java [hunk 1]
```diff
@@ -74,29 +81,12 @@ final class FetchSearchPhase extends SearchPhase {
...
-            @Override
-            public void onFailure(Exception e) {
-                context.onPhaseFailure(FetchSearchPhase.this, "", e);
-            }
-        });
-    }
-
-    private void innerRun() throws Exception {
         final int numShards = context.getNumShards();
-        final SearchPhaseController.ReducedQueryPhase reducedQueryPhase = resultConsumer.reduce();
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/action/search/FetchSearchPhase.java:99]: this change means the `resultConsumer` is no longer used at all in the fetch phase, could you remove it from the class?

File: server/src/main/java/org/elasticsearch/action/search/RankFeaturePhase.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+/**
+ * This search phase is responsible for executing any re-ranking needed for the given search request, iff that is applicable.
+ * It starts by retrieving {code num_shards * window_size} results from the query phase and reduces them to a global list of
+ * the top {@code window_size} results. It then reaches out to the shards to extract the needed feature data,
+ * and finally passes all this information to the appropriate {@code RankFeatureRankCoordinatorContext} which is responsible for reranking
+ * the results. If no rank query is specified, it proceeds directly to the next phase (FetchSearchPhase) by first reducing the results.
+ */
+public final class RankFeaturePhase extends SearchPhase {
+
+    private static final Logger logger = LogManager.getLogger(RankFeaturePhase.class);
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/action/search/RankFeaturePhase.java:25]: Neither the logger nor progress listener are used, maybe remove?

**Review by benwtrent — COMMENTED** [review]
Some minor comments. The changes seem sane to me. I read through the fetch phase to see if anything else needed to be forked, but really the only thing executing and doing any work on the calling thread was that reduce. All the work is then sent out to the shards to complete.

File: server/src/main/java/org/elasticsearch/action/search/RankFeaturePhase.java (hunk 2 above)
Comment (pmpailis) [reply server/src/main/java/org/elasticsearch/action/search/RankFeaturePhase.java:25]: Copied over from the WIP branch; removed them for now and will add them again with the phase impl PR

File: server/src/main/java/org/elasticsearch/action/search/FetchSearchPhase.java (hunk 1 above)
Comment (pmpailis) [reply server/src/main/java/org/elasticsearch/action/search/FetchSearchPhase.java:99]: Removed.

**Comment by pmpailis** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/action/search/FetchSearchPhase.java [hunk 3]
```diff
@@ -72,31 +78,13 @@ final class FetchSearchPhase extends SearchPhase {
...
     }
 
     @Override
     public void run() {
-        context.execute(new AbstractRunnable() {
-            @Override
-            protected void doRun() throws Exception {
-                // we do the heavy lifting in this inner run method where we reduce aggs etc. that's why we fork this phase
-                // off immediately instead of forking when we send back the response to the user since there we only need
-                // to merge together the fetched results which is a linear operation.
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/action/search/FetchSearchPhase.java:86]: While this comment implies that this is the ONLY reason for forking to the context executable, this comment was made 7 years ago. I would be more comfortable if you added back the fork to the `execute` here. 

I suspect some other things have snuck into the fetch phase that we wouldn't want executing on the same thread as the caller.

**Comment by pmpailis** [comment]
The renaming for the `*RankContext` classes has now been merged (https://github.com/elastic/elasticsearch/pull/107093).

@benwtrent could you please take a look and advise if any more changes are needed here?

**Review by benwtrent — APPROVED** [review]
:godmode:

**Comment by pmpailis** [comment]
@elasticmachine update branch

**Pull request merged**